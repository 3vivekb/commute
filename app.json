{
  "name": "commute",
  "scripts": {
  },
  "env": {
    "AUTH0_SIGNING_CERTIFICATE": {
      "required": true
    },
    "MAPZEN_SEARCH_KEY": {
      "required": true
    },
    "MONGODB_URI": {
      "required": true
    },
    "R5_URL": {
      "required": true
    },
    "STATIC_HOST": {
      "required": true
    }
  },
  "formation": {
  },
  "addons": [
    "mongolab"
  ],
  "buildpacks": [
    {
      "url": "heroku/nodejs"
    }
  ]
}
