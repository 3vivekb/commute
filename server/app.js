const bodyParser = require('body-parser')
const express = require('express')
const jwt = require('express-jwt')
const morgan = require('morgan')
const path = require('path')
const html = require('@conveyal/woonerf/html')

const env = require('./utils/env').env
const routes = require('./routes')

const app = express()

// middleware
let jwtMiddleWare
if (env.AUTH0_SIGNING_CERTIFICATE && process.env.NODE_ENV !== 'test') {
  jwtMiddleWare = jwt({
    algorithms: ['HS256', 'RS256'],
    secret: env.AUTH0_SIGNING_CERTIFICATE
  })
} else {
  jwtMiddleWare = (req, res, next) => {
    req.user = {
      sub: 'test-user'
    }
    next()
  }
}
app.use(morgan('combined'))
app.use(bodyParser.json({ limit: '50mb' }))

// static assets
app.use('/assets', express.static(path.resolve(__dirname, '../assets')))

// api
routes(app, jwtMiddleWare)

const htmlString = html({
  staticHost: process.env.STATIC_HOST || '/',
  title: 'Commute'
})

// webapp
app.get('*', (req, res) => {
  res.status(200).type('html').send(htmlString)
})

module.exports = app
