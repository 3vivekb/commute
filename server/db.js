const mongoose = require('mongoose')

mongoose.Promise = global.Promise // default is mpromise...ugh

const URI = process.env.MONGODB_URI || `mongodb://localhost/commute${process.env.test ? '-test' : ''}`
const db = module.exports = mongoose.createConnection(URI)

db.on('connected', () => console.log('Mongoose default connection open to ' + URI))
db.on('error', (err) => console.log('Mongoose default connection error: ' + err))
db.on('disconnected', () => console.log('Mongoose default connection disconnected'))

// If the Node process ends, close the Mongoose connection
process.once('SIGINT', () =>
  db.close(() =>
    console.log('Mongoose default connection disconnected through app termination') && process.exit(0)))

// For NODEMON
process.once('SIGUSR2', () =>
  db.close(() =>
    console.log('Mongoose default connection disconnected through nodemon restart') && process.kill(process.pid, 'SIGUSR2')))
