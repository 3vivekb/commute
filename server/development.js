const app = require('./app')
const db = require('./db')

const PORT = process.env.PORT || 9966

db.on('connected', () =>
  app.listen(PORT, (err) =>
    err ? console.error(err) : console.log('GraphQL API now running on port', PORT)))
