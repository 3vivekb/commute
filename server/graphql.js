const {buildSchema} = require('graphql')

const {
  Organization
} = require('./models')

const schema = buildSchema(`
  input Find {
    _id: ID
    owner: String
  }

  type Point {
    type: String!
    coordinates: [Float]
  }

  type Commuter {
    _id: ID!
    name: String!
    location: Point
  }

  type Group {
    _id: ID!
    name: String!
    organization: Organization!
    commuters: [Commuter]
  }

  type Organization {
    _id: ID!
    owner: String!
    name: String!
    update(name: String!): Organization
  }

  type Site {
    _id: ID!
    location: Point!
    organization: Organization!
  }

  type SiteAnalysis {
    _id: ID!
    group: Group!
    organization: Organization!
    site: Site!
  }

  type Query {
    organization(find: Find): Organization
    organizations: [Organization]
  }

  type Mutation {
    createOrganization(name: String!): Organization
    deleteOrganization(name: String!): Organization
    organization(_id: ID!): Organization
    updateOrganization(name: String! newName: String!): Organization
  }
`)

const rootValue = {
  createOrganization (properties, request) {
    const o = new Organization({
      owner: request.user || 'Trevor',
      name: properties.name
    })
    return o.save()
  },
  deleteOrganization (properties, request) {
    return Organization
      .findOne({ name: properties.name })
      .exec((res) => {
        return Organization.remove(res).exec()
      })
  },
  organization ({find}, request) {
    return Organization.findOne(find).exec()
  },
  organizations (properties, request) {
    return Organization
      .find()
      .exec()
  },
  updateOrganization (properties, request) {
    return Organization.findOneAndUpdate({
      name: properties.name
    }, {
      name: properties.newName
    }).exec()
  }
}

module.exports = {schema, rootValue}
