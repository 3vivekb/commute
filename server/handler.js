const {graphql} = require('graphql')
const {schema, root} = require('./graphql')

module.exports.graphql = (event, context, callback) =>
  graphql(schema, event.body.query, root, {}, event.body.variables)
    .then((response) => callback(null, response))
    .catch((error) => callback(error))
