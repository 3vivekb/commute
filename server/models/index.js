const db = require('../db')

module.exports = {
  Commuter: db.model('Commuter', require('./commuter')),
  Group: db.model('Group', require('./group')),
  Organization: db.model('Organization', require('./organization')),
  Analysis: db.model('Analysis', require('./analysis')),
  Site: db.model('Site', require('./site'))
}
