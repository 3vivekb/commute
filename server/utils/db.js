const each = require('async/each')

export function makeCascadeDeleteModelsFn (foreignKey, models) {
  return function (next, done) {
    // CASCADE DELETE if needed
    if (this.isModified('trashed') && this.trashed) {
      const foreignKeyQuery = {}
      foreignKeyQuery[foreignKey] = this._id
      each(models, (model) => {
        model.find(foreignKeyQuery, (err, docs) => {
          if (err) return done(err)
          each(docs, (doc, cb) => {
            doc.trash(cb)
          }, done)
        })
      }, done)
    }
    next()
  }
}
