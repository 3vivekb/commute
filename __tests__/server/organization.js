/* globals afterAll, beforeAll, describe, expect, it */

import mongoose from 'mongoose'
import request from 'supertest-as-promised'

import app from '../../server/app'
import db from '../../server/db'
import {Organization} from '../../server/models'

describe('organization', () => {
  beforeAll((done) => {
    // start with fresh db
    db.dropDatabase(done)
  })

  afterAll(() => {
    mongoose.disconnect() // disconnect from mongo to end running of tests
  })

  it('should find zero organizations in fresh state', () => {
    return request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({
        query: `{
          organizations{
            name
          }
        }`
      }))
      .then((res) => {
        const json = JSON.parse(res.text)
        expect(json.errors).toBeFalsy()
        expect(json.data.organizations.length).toBe(0)
      })
  })

  it('should add organization', () => {
    return request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({
        query: `mutation Mutation {
          createOrganization(name: "test-org") {
            name
          }
        }`
      }))
      .then(async (res) => {
        const json = JSON.parse(res.text)
        expect(json.errors).toBeFalsy()
        expect(json).toMatchSnapshot()
        const count = await Organization.count().exec()
        expect(count).toBe(1)
      })
  })

  it('should find organization after it was added', () => {
    return request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({
        query: `{
          organizations{
            owner,
            name
          }
        }`
      }))
      .then((res) => {
        const json = JSON.parse(res.text)
        expect(json.errors).toBeFalsy()
        expect(json.data.organizations.length).toBe(1)
        expect(json.data.organizations[0].name).toBe('test-org')
      })
  })

  it('should update organization', () => {
    return request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({
        query: `mutation Mutation {
          updateOrganization(
            name: "test-org"
            newName: "New Organization Name"
          ) {
            name
          }
        }`
      }))
      .then(async (res) => {
        const json = JSON.parse(res.text)
        expect(json.errors).toBeFalsy()
        const org = await Organization.findOne({ name: 'New Organization Name' }).exec()
        expect(org.name).toBe('New Organization Name')
      })
  })

  it('should delete organization', () => {
    return request(app)
      .post('/graphql')
      .set('Content-Type', 'application/json')
      .send(JSON.stringify({
        query: `mutation Mutation {
          deleteOrganization(
            name: "New Organization Name"
          ) {
            name
          }
        }`
      }))
      .then(async (res) => {
        const json = JSON.parse(res.text)
        expect(json.errors).toBeFalsy()
        const count = await Organization.count().exec()
        expect(count).toBe(0)
      })
  })
})
