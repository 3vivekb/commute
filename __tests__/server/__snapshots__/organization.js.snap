exports[`organization should add organization 1`] = `
Object {
  "data": Object {
    "createOrganization": Object {
      "name": "test-org",
    },
  },
}
`;
